vllm/core/block_manager_v1.py

```diff
@@ -301,10 +315,44 @@ def allocate(self, seq_group: SequenceGroup) -> None:
                 block.ref_count = seq_group.num_seqs()
             block_table.append(block)
 
-        # Assign the block table for each sequence.
+        # Assign the decoder block table for each sequence.
         for seq in seq_group.get_seqs(status=SequenceStatus.WAITING):
             self.block_tables[seq.seq_id] = block_table.copy()
 
+    def allocate_cross_block_table(self, seq_group: SequenceGroup) -> None:
+        # NOTE: Here we assume that all sequences in the group have the same
+        # encoder prompt.
+
+        # Allocate new physical token blocks that will store the prompt tokens.
+        seq = seq_group.get_encoder_seq()
+        if seq is not None:
+            block_table: BlockTable = []
+            num_prompt_blocks = len(seq.logical_token_blocks)
+            for logical_idx in range(num_prompt_blocks):
+                if (self.block_sliding_window is not None
+                        and logical_idx >= self.block_sliding_window):
+                    block = block_table[logical_idx %
+                                        self.block_sliding_window]
+                    # Set the reference counts of the token blocks.
+                    block.ref_count = seq_group.num_seqs()
+                elif self.enable_caching:
```

Comment:
Let me revisit this, then I will clarify

Reply:
I am assuming that encoder attention is usually self attention. Hence if this is prefix cache support I wonder how much it would be useful for cross attention given that it is self attention. This might only be useful if it is an exact match?

Reply:
Yes this is true. In fact, prefix caching support is not part of this workstream, I am not holding myself accountable to support prefix caching functionality alongside encoder/decoder regardless.

Given that that line of code is part of a larger block that I just refactored into a single helper function, and given that that helper function will be shared between decoder-only and encoder/decoder models, (and assuming that enable_caching *is* actually related to prefix caching), perhaps I will simply modify the if statement so that the enable_caching branch of the if statement cannot run for an encoder/decoder model.

Reply:
Confirmed that enable_caching is the automatic prefix caching flag; just pushed a commit which enforces that encoder/decoder does not currently support automatic prefix caching.